lib/pure/collections/sequtils.nim

```diff
@@ -774,48 +781,86 @@ template anyIt*(s, pred: untyped): bool =
       break
   result
 
-template toSeq1(s: not iterator): untyped =
-  # overload for typed but not iterator
-  type OutType = typeof(items(s))
-  when compiles(s.len):
-    block:
-      evalOnceAs(s2, s, compiles((let _ = s)))
-      var i = 0
-      var result = newSeq[OutType](s2.len)
-      for it in s2:
-        result[i] = it
-        i += 1
-      result
-  else:
-    var result: seq[OutType] = @[]
-    for it in s:
-      result.add(it)
-    result
+template toSeqBuiltin[IX, T](arg: array[IX, T]): seq[T] = @arg
 
-template toSeq2(iter: iterator): untyped =
-  # overload for iterator
-  evalOnceAs(iter2, iter(), false)
-  when compiles(iter2.len):
-    var i = 0
-    var result = newSeq[typeof(iter2)](iter2.len)
-    for x in iter2:
-      result[i] = x
-      inc i
-    result
+template toSeqBuiltin[T](arg: openArray[T]): seq[T] = @arg
+
+template toSeqBuiltin(arg: string): seq[char] = @arg
+
+template toSeqBuiltin[T](arg: seq[T]): seq[T] = arg
+
+proc toSeqBuiltin[T](arg: set[T]): seq[T] {.inline.} =
+  for it in arg:
+    result.add it
+
+proc toSeqBuiltin(a: cstring): seq[char] {.inline.} =
+  let len = a.len
+  result.setLen len
+  var i = 0
+  while i < len:
+    result[i] = a[i]
+    inc i
+
+proc toSeqBuiltin[E: enum](t: typedesc[E]): seq[E] =
+  for it in t:
+    result.add it
+
+macro toSeqBuiltin(arg: iterator): untyped =
+  let typ = arg.getTypeInst
+  typ.expectKind nnkProcTy # only for closure iterators
+  let formalParams = typ[0]
+  formalParams.expectKind nnkFormalParams
+  formalParams.expectMinLen 1
+  let resultTyp = nnkBracketExpr.newTree(bindSym"seq", formalParams[0])
```

Comment:
Wouldn't this explode spectacularly if generics are used?

Reply:
can you provide an example? I am asking, because also for me it was quite some time age since I originally wrote this rewrite of `toSeq`, it is just recent that I contributed it to nimskull.

Reply:
I couldn't get that part to trip up, which is great, but I caused some other code to fail, though.